daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala

```diff
@@ -223,41 +223,69 @@ private[lf] object Speedy {
 
   @throws[SErrorDamlException]
   private[speedy] def buildDiscTable(
-      disclosures: ImmArray[speedy.DisclosedContract]
+      disclosures: ImmArray[speedy.DisclosedContract],
+      compiledPackages: CompiledPackages,
   ): DisclosureTable = {
     val _ = disclosures
     val acc = disclosures.foldLeft(
-      (
-        DisclosureTable.Empty
-      )
+      DisclosureTable.Empty
     ) { case (table, d) =>
       val arg = d.argument
       val coid = d.contractId
       // check for duplicate contract ids
       table.contractById.get(coid) match {
         case Some(_) =>
-          throw (SErrorDamlException(
+          throw SErrorDamlException(
             IError.DisclosurePreprocessing(
               IError.DisclosurePreprocessing.DuplicateContractIds(d.templateId)
             )
-          ))
+          )
+
         case None =>
           val m1_prime = table.contractById + (coid -> (d.templateId, arg))
           d.metadata.keyHash match {
             case Some(hash) =>
               // check for duplicate contract key hashes
               table.contractIdByKey.get(hash) match {
                 case Some(_) =>
-                  throw (SErrorDamlException(
+                  throw SErrorDamlException(
                     IError.DisclosurePreprocessing(
                       IError.DisclosurePreprocessing.DuplicateContractKeys(
                         d.templateId
                       )
                     )
-                  ))
+                  )
+
                 case None => DisclosureTable(table.contractIdByKey + (hash -> coid), m1_prime)
               }
-            case None => table.copy(contractById = m1_prime)
+
+            case None =>
+              compiledPackages.pkgInterface.lookupTemplate(d.templateId) match {
+                case Right(template) if template.key.isEmpty =>
+                  // Success - template exists, but has no key defined
+                  table.copy(contractById = m1_prime)
+
+                case Right(_) =>
+                  // Error - template exists, but has a key defined
```

Comment:
why is it an error for the contract to have a key?

Reply:
In this context, the code knows that the disclosed contract metadata has no key hash. When we now go and check the template associated with the disclosed contract ID, we discover a key is defined on this template.

Now, @cocreature has commented as follows:

> This seems invalid: If the template has a key, we should mandate the key not silently accept that people don’t pass the hash. I think we should:
> 
> Add a test that this gets rejected if you don’t pass a hash and the template defines a key.
> Change these tests to use different templates for with and without key.

As the template has a key associated with it, when the disclosed contract instance is defined, then we should have captured the key hash. Not having the key hash (in the disclosed contract metadata) when the contract template has a key is thus an error scenario.

Reply:
ah, thanks for the explanation, I understand now. 

Perhaps the error message should be reworded to emphasize that the actionable problem is the lack of a key hash (not the presence of a key)

> Template has a key defined, but there is no key hash for disclosed contract

\>>>

> Disclosed contract lacks a key hash, but the template requires a key

I'm not too invested in the color of the bikeshed, though.

(I wonder, though, if we could have an API that _just_ does the right thing, i.e. adds the key hash for templates with a key using the template's key expression, and doesn't add it for templates without one. In pure Haskell it's usually a pain to write code conditional on the existence of an instance (`HasKey` in this case), but a primitive/builtin would do the trick in Daml)

Reply:
@akrmn I'm not quite sure what API change you're suggesting. These are passed in over the ledger API. I don't think we can structure those protobufs such that we enforce that they're "correct by construction". But I'm also not super worried about this. The usage we expect is:

1. You read the contract from the ledger API which includes the hash iff he template has a key.
2. You pass it in for explicit disclosure including the full metadata.

So unless you manually start messing around with the value you get from the ledger API (which you should not), things should just work™.

Reply:
@cocreature thanks for explaining. I was thinking in terms of an old PR from Robin where we had step 2 but not step 1, so I wasn't sure how one should construct the metadata, and I thought we could provide a function `template -> DisclosureMetadata` that did the right thing - now it's clear that one should just not construct the metadata at all 😄.